src/dca.jl

```diff
@@ -0,0 +1,596 @@
+                @info "Timeout reached after $elapsed_time seconds."
+            end
+            break
+        end
+
+        # Step 1: Compute gradient of concave part at current point
+        # ∇g(x_t) - this remains constant throughout the subproblem
+        grad_g!(grad_g_workspace, x_current)
+
+        # Step 2: Precompute constants for linearized objective m(x)
+        # m(x) = f(x) - g(x_t) - ⟨∇g(x_t), x - x_t⟩  
+        #      = f(x) - g(x_t) - ⟨∇g(x_t), x⟩ + ⟨∇g(x_t), x_t⟩
+        g_value_at_current = g(x_current)
+        grad_g_dot_current = dot(grad_g_workspace, x_current)
+
+        # Create closures for the linearized subproblem
+        linearized_objective =
+            x -> _dca_linearized_objective(
+                x,
+                f,
+                g_value_at_current,
+                grad_g_workspace,
+                grad_g_dot_current,
+            )
+        linearized_gradient! =
+            (storage, x) -> _dca_linearized_gradient!(storage, x, grad_f!, grad_g_workspace)
+
+        # Step 3: Setup early stopping callback if enabled
+        inner_callback = nothing
+        if use_dca_early_stopping
+            phi_value_at_current = f(x_current) - g(x_current)
+            inner_callback = make_dca_early_stopping_callback(
+                nothing,
+                linearized_objective,
+                linearized_gradient!,
+                phi_value_at_current,
+            )
+        end
+
+        # Step 4: Solve the convex subproblem min_x m(x) using Frank-Wolfe variants
+        x_inner_start = copy(x_current)
+
+        if !use_corrective_fw
+            # Use standard Frank-Wolfe
+            fw_result = frank_wolfe(
+                linearized_objective,
+                linearized_gradient!,
+                lmo,
+                x_inner_start,
+                line_search=line_search,
+                epsilon=epsilon * DEFAULT_DCA_EPSILON_FACTOR,
+                max_iteration=max_inner_iteration,
+                trajectory=true,
+                verbose=verbose_inner,
+                print_iter=print_iter_inner,
+                memory_mode=memory_mode,
+                linesearch_workspace=linesearch_workspace,
+                timeout=timeout - elapsed_time,
+                callback=inner_callback,
+            )
+        else # Use an active-set based FW variant
+            # if active set not created already, or if we don't warm-start
+            # create an active set from scratch
+            if active_set === nothing || !warm_start
+                active_set = ActiveSet([1.0], [x0], similar(x0))
+            end
+            fw_result = corrective_frank_wolfe(
+                linearized_objective,
+                linearized_gradient!,
+                lmo,
+                corrective_subsolver,
+                active_set,
+                line_search=line_search,
+                epsilon=epsilon * DEFAULT_DCA_EPSILON_FACTOR,
+                max_iteration=max_inner_iteration,
```

Comment:
i think when you do the "second" round, the active_set used it the current one and that also changes the point needed for computation of the dca gap. hence the negative values.